diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index 6031ec58c7ae..44bcbf060718 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -4731,9 +4731,6 @@ static void __setscheduler_params(struct task_struct *p,
        p->rt_priority = attr->sched_priority;
        p->normal_prio = normal_prio(p);
        set_load_weight(p, true);
-
-       if (attr->sched_flags & SCHED_FLAG_LATENCY_NICE)
-               p->latency_nice = attr->sched_latency_nice;
 }

 /* Actually do priority change: must hold pi & rq lock. */
@@ -4749,6 +4746,13 @@ static void __setscheduler(struct rq *rq, struct
task_struct *p,

        __setscheduler_params(p, attr);

+       /*
+        * Change latency_nice value only when SCHED_FLAG_LATENCY_NICE or
+        * SCHED_FLAG_ALL sched_flag is set.
+        */
+       if (attr->sched_flags & SCHED_FLAG_LATENCY_NICE)
+               p->latency_nice = attr->sched_latency_nice;
+

This should allow setting value only on above flags, also restricts setting
the value when SCHED_FLAG_KEEP_PARAMS/SCHED_FLAG_KEEP_ALL is passed.


Thanks,
Parth

  