---
 block/elevator.c | 6 +++++-
 init/Kconfig     | 4 ++++
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/block/elevator.c b/block/elevator.c
index 0a55c7aa9249f..214ca53eed6ca 100644
--- a/block/elevator.c
+++ b/block/elevator.c
@@ -628,10 +628,14 @@ static inline bool elv_support_iosched(struct request_queue *q)
  */
 static struct elevator_type *elevator_get_default(struct request_queue *q)
 {
+#ifndef CONFIG_ZEN_INTERACTIVE
 	if (q->nr_hw_queues != 1)
 		return NULL;
+#endif
 
-#if defined(CONFIG_MQ_IOSCHED_DEADLINE_NODEFAULT)
+#if defined(CONFIG_ZEN_INTERACTIVE) && defined(CONFIG_MQ_IOSCHED_I10)
+	return elevator_get(q, "i10", false);
+#elif defined(CONFIG_MQ_IOSCHED_DEADLINE_NODEFAULT)
 	return elevator_get(q, "mq-deadline-nodefault", false);
 #else
 	return elevator_get(q, "mq-deadline", false);
diff --git a/init/Kconfig b/init/Kconfig
index f8f25c21f2a8c..b79e2c2232573 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -98,6 +98,10 @@ config ZEN_INTERACTIVE
 	help
 	  Tunes the kernel for responsiveness at the cost of throughput and power usage.
 
+	  --- Block Layer ----------------------------------------
+
+	    Default scheduler.........: mq-deadline -> kyber
+
 config BROKEN
 	bool
 
