From cb8ba8c84837eb6921ace9d1ec831a64a70f69c7 Mon Sep 17 00:00:00 2001
From: "Jan Alexander Steffens (heftig)" <jan.steffens@gmail.com>
Date: Mon, 27 Jan 2020 18:19:11 +0100
Subject: [PATCH] ZEN: INTERACTIVE: Increase default writeback thresholds

---
 init/Kconfig        | 5 +++++
 mm/page-writeback.c | 8 ++++++++
 2 files changed, 13 insertions(+)

diff --git a/init/Kconfig b/init/Kconfig
index 4720f0c29b8c4..4881144228145 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -102,6 +102,11 @@ config ZEN_INTERACTIVE
 
 	    Default scheduler.........: mq-deadline -> bfq
 
+	  --- Virtual Memory Subsystem ---------------------------
+
+	    Mem dirty before bg writeback..:  10 %  ->  20 %
+	    Mem dirty before sync writeback:  20 %  ->  50 %
+
 config BROKEN
 	bool
 
diff --git a/mm/page-writeback.c b/mm/page-writeback.c
index 28b3e7a675657..306a90e881d6c 100644
--- a/mm/page-writeback.c
+++ b/mm/page-writeback.c
@@ -71,7 +71,11 @@ static long ratelimit_pages = 32;
 /*
  * Start background writeback (via writeback threads) at this percentage
  */
+#ifdef CONFIG_ZEN_INTERACTIVE
+int dirty_background_ratio = 20;
+#else
 int dirty_background_ratio = 10;
+#endif
 
 /*
  * dirty_background_bytes starts at 0 (disabled) so that it is a function of
@@ -88,7 +92,11 @@ int vm_highmem_is_dirtyable;
 /*
  * The generator of dirty data starts writeback at this percentage
  */
+#ifdef CONFIG_ZEN_INTERACTIVE
+int vm_dirty_ratio = 50;
+#else
 int vm_dirty_ratio = 20;
+#endif
 
 /*
  * vm_dirty_bytes starts at 0 (disabled) so that it is a function of
